File: app/src/main/java/education/mahmoud/quranyapp/feature/ayahs_search/ShowSearchResults.java
package education.mahmoud.quranyapp.feature.ayahs_search;

import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.flipboard.bottomsheet.BottomSheetLayout;
import com.flipboard.bottomsheet.commons.MenuSheetView;
import com.google.android.material.textfield.TextInputEditText;

import org.koin.java.KoinJavaComponent;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import education.mahmoud.quranyapp.R;
import education.mahmoud.quranyapp.datalayer.Repository;
import education.mahmoud.quranyapp.datalayer.local.room.AyahItem;
import education.mahmoud.quranyapp.feature.listening_activity.AyahsListen;
import education.mahmoud.quranyapp.feature.listening_activity.ListenServie;
import education.mahmoud.quranyapp.feature.show_sura_ayas.ShowAyahsActivity;
import education.mahmoud.quranyapp.utils.Constants;
import education.mahmoud.quranyapp.utils.Util;

public class ShowSearchResults extends AppCompatActivity {

    private static final String TAG = "ShowSearchResults";
    private static final int RC_STORAGE = 10002;
    @BindView(R.id.edSearch)
    TextInputEditText edSearch;
    @BindView(R.id.rvSearch)
    RecyclerView rvSearch;
    @BindView(R.id.tvNotFound)
    TextView tvNotFound;
    @BindView(R.id.tvSearchCount)
    TextView tvSearchCount;

    SearchResultsAdapter adapter;
    @BindView(R.id.bottomSearch)
    BottomSheetLayout bottomSearch;
    private Repository repository = KoinJavaComponent.get(Repository.class);
    private List<AyahItem> ayahItems;
    private String ayah;

    boolean isRunning = false;

    private void editWatcher() {
        edSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                ayah = editable.toString();
                ayah = ayah.replaceAll(" +", " ");
                if (!TextUtils.isEmpty(ayah) && ayah.length() > 0) {
                    ayahItems = repository.getAyahByAyahText(ayah);
                    Log.d(TAG, "afterTextChanged: " + ayahItems.size());

                    int count = ayahItems.size(); // n of results
                    tvSearchCount.setText(getString(R.string.times, count));
                    if (count > 0) {
                        adapter.setAyahItemList(ayahItems, ayah);
                        foundState();
                    } else {
                        notFoundState();
                    }
                } else {
                    defaultState();
                }

            }
        });

    }

    private void setUpBottomSheet(AyahItem ayahItem) {
        // bottom sheet
        MenuSheetView menuSheetView =
                new MenuSheetView(ShowSearchResults.this, MenuSheetView.MenuType.LIST, "Options", new MenuSheetView.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem item) {
                        if (bottomSearch.isSheetShowing()) {
                            bottomSearch.dismissSheet();
                        }
                        switch (item.getItemId()) {
                            case R.id.menuOpen:
                                openPage(ayahItem.getPageNum());
                                break;
                            case R.id.menuPlaySound:
                                playAudio(ayahItem);
                                break;
                            case R.id.menuTafser:
                                showTafseer(ayahItem);
                                break;
                        }
                        return true;
                    }
                });
        menuSheetView.inflateMenu(R.menu.menu_sheet_search);
        bottomSearch.showWithSheetView(menuSheetView);
    }

    private void showTafseer(AyahItem ayahItem) {
        if (ayahItem.getTafseer() != null) {
            String title = this.
                    getString(R.string.tafserr_info, ayahItem.getAyahInSurahIndex(), ayahItem.getPageNum(), ayahItem.getJuz());
            Util.getDialog(this, ayahItem.getTafseer(), title).show();
        } else {
            Toast.makeText(this, this.getText(R.string.tafseer_not_down), Toast.LENGTH_SHORT).show();
        }
    }

    private void adapterListeners() {
        adapter.setiSearchItemClick(new SearchResultsAdapter.ISearchItemClick() {
            @Override
            public void onSearchItemClick(AyahItem item) {
                setUpBottomSheet(item);
                Log.d(TAG, "onSearchItemClick: ");
            }
        });
    }

    private void openPage(int pageNum) {
        Intent openAcivity = new Intent(ShowSearchResults.this, ShowAyahsActivity.class);
        openAcivity.putExtra(Constants.PAGE_INDEX, pageNum);
        startActivity(openAcivity);
    }

    /**
     * state with no search performed
     */
    private void defaultState() {
        tvNotFound.setVisibility(View.GONE);
    }

    private void initRv() {
        LinearLayoutManager manager = new LinearLayoutManager(this);
        rvSearch.setLayoutManager(manager);

        Typeface typeface = Typeface.createFromAsset(getAssets(), "font/me_quran.ttf");
        adapter = new SearchResultsAdapter(typeface);
        rvSearch.setAdapter(adapter);
        rvSearch.setHasFixedSize(true);

    }

    private void foundState() {
        rvSearch.setVisibility(View.VISIBLE);
        tvNotFound.setVisibility(View.GONE);
    }

    private void notFoundState() {
        rvSearch.setVisibility(View.GONE);
        tvNotFound.setVisibility(View.VISIBLE);
    }

    Intent serviceIntent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_search_results);
        ButterKnife.bind(this);

        initRv();
        adapterListeners();
        editWatcher();

    }

    private void playAudio(AyahItem item) {
        if (item.getAudioPath() != null) {
            Log.d(TAG, "playAudio: isRunning  " + isRunning);
            List<AyahItem> ayahItems = new ArrayList<>();
            ayahItems.add(item);

            AyahsListen ayahsListen = new AyahsListen();
            ayahsListen.setAyahItemList(ayahItems);

            if (serviceIntent != null) {
                stopService(serviceIntent);
            }
            serviceIntent = ListenServie.createService(getApplicationContext(),
                    ayahsListen );

        } else {
            showMessage(getString(R.string.not_downlod_audio));
        }

    }

    private void showMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/education/mahmoud/quranyapp/feature/ayahs_search/ShowSearchResults.kt
package education.mahmoud.quranyapp.feature.ayahs_search

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.flipboard.bottomsheet.commons.MenuSheetView
import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
import education.mahmoud.quranyapp.R
import education.mahmoud.quranyapp.base.BaseFragment
import education.mahmoud.quranyapp.databinding.ActivityShowSearchResultsBinding
import education.mahmoud.quranyapp.databinding.HeaderSearchBinding
import education.mahmoud.quranyapp.datalayer.QuranRepository
import education.mahmoud.quranyapp.datalayer.local.room.AyahItem
import education.mahmoud.quranyapp.feature.listening_activity.AyahsListen
import education.mahmoud.quranyapp.feature.listening_activity.ListenServie
import education.mahmoud.quranyapp.feature.showSuraAyas.ShowAyahsActivity
import education.mahmoud.quranyapp.utils.Constants
import education.mahmoud.quranyapp.utils.Util
import org.koin.java.KoinJavaComponent
import java.util.* // ktlint-disable no-wildcard-imports

class ShowSearchResults : BaseFragment() {

    private var adapter = SearchResultsAdapter()

    private val repository = KoinJavaComponent.get(QuranRepository::class.java)
    private var ayahItems: List<AyahItem> = listOf()
    private var ayah: String = ""
    var isRunning = false

    private val binding by viewBinding(ActivityShowSearchResultsBinding::bind)
    private val searchBinding by viewBinding(HeaderSearchBinding::bind)

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        return inflater.inflate(R.layout.activity_show_search_results, container, false)
    }

    override fun initViews(view: View) {
        super.initViews(view)
        initRv()
        adapterListeners()
        editWatcher()
        setClickListeners()
    }

    override fun setClickListeners() {
        super.setClickListeners()

        searchBinding.imSearch.setOnClickListener {
            doSearch("${searchBinding.edSearch.text}")
        }
    }

    private fun editWatcher() {
        searchBinding.edSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                doSearch(editable.toString())
            }
        })
    }

    private fun doSearch(text: String?) {
        ayah = text ?: ""
        ayah = ayah.replace(" +".toRegex(), " ")
        if (ayah.isNotEmpty()) {
            ayahItems = repository.getAyahByAyahText(ayah)
            Log.d(TAG, "afterTextChanged: " + ayahItems.size)
            val count = ayahItems.size // n of results
            binding.tvSearchCount.text = getString(R.string.times, count)
            if (count > 0) {
                adapter.setAyahItemList(ayahItems, ayah)
                foundState()
            } else {
                notFoundState()
            }
        } else {
            defaultState()
        }
    }

    private fun setUpBottomSheet(ayahItem: AyahItem) { // bottom sheet
        val menuSheetView = MenuSheetView(
            requireContext(), MenuSheetView.MenuType.LIST, "Options",
            MenuSheetView.OnMenuItemClickListener { item ->
                if (binding.bottomSearch.isSheetShowing) {
                    binding.bottomSearch.dismissSheet()
                }
                when (item.itemId) {
                    R.id.menuOpen -> openPage(ayahItem.pageNum)
                    R.id.menuPlaySound -> playAudio(ayahItem)
                    R.id.menuTafser -> showTafseer(ayahItem)
                }
                true
            }
        )
        menuSheetView.inflateMenu(R.menu.menu_sheet_search)
        binding.bottomSearch.showWithSheetView(menuSheetView)
    }

    private fun showTafseer(ayahItem: AyahItem) {
        val title = this.getString(
            R.string.tafserr_info,
            ayahItem.ayahInSurahIndex,
            ayahItem.pageNum,
            ayahItem.juz
        )
        Util.getDialog(requireContext(), ayahItem.tafseer, title).show()
    }

    private fun adapterListeners() {
        adapter.setiSearchItemClick(object : SearchResultsAdapter.ISearchItemClick {
            override fun onSearchItemClick(item: AyahItem?) {
                item?.let { setUpBottomSheet(item) }
                Log.d(TAG, "onSearchItemClick: ")
            }
        })
    }

    private fun openPage(pageNum: Int) {
        activity?.let {
            val openAcivity = Intent(requireContext(), ShowAyahsActivity::class.java)
            openAcivity.putExtra(Constants.PAGE_INDEX, pageNum)
            startActivity(openAcivity)
        }
    }

    /**
     * state with no search performed
     */
    private fun defaultState() {
        binding.tvNotFound.visibility = View.GONE
    }

    private fun initRv() {
        binding.rvSearch.adapter = adapter
        binding.rvSearch.setHasFixedSize(true)
    }

    private fun foundState() {
        binding.rvSearch.visibility = View.VISIBLE
        binding.tvNotFound.visibility = View.GONE
    }

    private fun notFoundState() {
        binding.rvSearch.visibility = View.GONE
        binding.tvNotFound.visibility = View.VISIBLE
    }

    var serviceIntent: Intent? = null

    private fun playAudio(item: AyahItem) {
        if (item.audioPath != null) {
            Log.d(TAG, "playAudio: isRunning  $isRunning")
            val ayahItems: MutableList<AyahItem> = ArrayList()
            ayahItems.add(item)
            val ayahsListen = AyahsListen()
            ayahsListen.ayahItemList = ayahItems
            if (serviceIntent != null) {
                requireContext().stopService(serviceIntent)
            }
            serviceIntent = ListenServie.createService(
                requireActivity().applicationContext,
                ayahsListen
            )
        } else {
            showMessage(getString(R.string.not_downlod_audio))
        }
    }

    private fun showMessage(message: String) {
        Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT).show()
    }

    companion object {
        private const val TAG = "ShowSearchResults"
        private const val RC_STORAGE = 10002
    }
}
